frame/support/src/storage/mod.rs

```diff
@@ -44,82 +47,108 @@ pub mod types;
 pub mod unhashed;
 pub mod weak_bounded_vec;
 
-#[cfg(all(feature = "std", any(test, debug_assertions)))]
-mod debug_helper {
-	use std::cell::RefCell;
+mod storage_layer_tracker {
+	const STORAGE_LAYER_KEY: &'static [u8] = b":storage_layer:";
+	type Layer = u8;
 
-	thread_local! {
-		static TRANSACTION_LEVEL: RefCell<u32> = RefCell::new(0);
+	pub fn get_storage_layer() -> Layer {
+		crate::storage::unhashed::get_or_default::<Layer>(STORAGE_LAYER_KEY)
 	}
 
-	pub fn require_transaction() {
-		let level = TRANSACTION_LEVEL.with(|v| *v.borrow());
-		if level == 0 {
-			panic!("Require transaction not called within with_transaction");
-		}
+	fn set_storage_layer(level: &Layer) {
+		crate::storage::unhashed::put::<Layer>(STORAGE_LAYER_KEY, level);
 	}
 
-	pub struct TransactionLevelGuard;
-
-	impl Drop for TransactionLevelGuard {
-		fn drop(&mut self) {
-			TRANSACTION_LEVEL.with(|v| *v.borrow_mut() -= 1);
-		}
+	fn kill_storage_layer() {
+		crate::storage::unhashed::kill(STORAGE_LAYER_KEY);
 	}
 
-	/// Increments the transaction level.
+	/// Increments the transaction level. Returns an error if levels go past the limit.
 	///
 	/// Returns a guard that when dropped decrements the transaction level automatically.
-	pub fn inc_transaction_level() -> TransactionLevelGuard {
-		TRANSACTION_LEVEL.with(|v| {
-			let mut val = v.borrow_mut();
-			*val += 1;
-			if *val > 10 {
-				log::warn!(
-					"Detected with_transaction with nest level {}. Nested usage of with_transaction is not recommended.",
-					*val
-				);
-			}
-		});
+	pub fn inc_storage_layer(limit: Layer) -> Result<StorageLayerGuard, ()> {
+		let existing_levels = get_storage_layer();
+		if existing_levels >= limit || existing_levels == Layer::MAX {
+			return Err(())
+		}
+		// Cannot overflow because of check above.
+		set_storage_layer(&(existing_levels + 1));
+		Ok(StorageLayerGuard)
+	}
 
-		TransactionLevelGuard
+	fn dec_storage_layer() {
+		let existing_levels = get_storage_layer();
+		if existing_levels == 0 {
+			log::warn!(
```

Comment:
Debug assert maybe?  
When we extend the `defensive_trait` from @kianenigma we could have a proper solution here.